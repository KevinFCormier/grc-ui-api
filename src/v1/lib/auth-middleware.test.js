/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import createAuthMiddleWare from './auth-middleware';

describe('Auth Middleware', () => {
  const mockCache = {
    set: jest.fn(),
    get: jest.fn(),
  };

  const mockHttpLib = jest.fn().mockReturnValue({ body: { id_token: 'test-id-token' } });

  test('Sets kubeToken to Bearer localdev in development', (done) => {
    const mockRequest = { headers: { authorization: '' } };
    const authMiddleware = createAuthMiddleWare({ shouldLocalAuth: true });

    authMiddleware(mockRequest, null, (err) => {
      expect(err).not.toBeDefined();
      expect(mockRequest.kubeToken).toBe('Bearer localdev');
      done();
    });
  });

  test('Caches token on first authorization', (done) => {
    const mockRequest = { headers: { authorization: 'substring-for-auth' } };
    const authMiddleware = createAuthMiddleWare({
      cache: mockCache,
      httpLib: mockHttpLib,
      shouldLocalAuth: false,
    });

    authMiddleware(mockRequest, null, (err) => {
      expect(err).not.toBeDefined();
      expect(mockRequest.kubeToken).toBe('Bearer test-id-token');
      expect(mockHttpLib.mock.calls).toHaveLength(1);
      expect(mockCache.set.mock.calls).toHaveLength(3);
      expect(mockCache.set.mock.calls[0]).toMatchSnapshot();
      done();
    });
  });

  test('Retrieves token from cache if present', (done) => {
    mockHttpLib.mockClear();
    mockCache.get.mockClear();

    const mockRequest = { headers: { authorization: 'substring-for-auth' } };
    mockCache.get.mockReturnValueOnce(new Promise(resolve => resolve({ body: { id_token: 'test-cached-token' } })));

    const authMiddleware = createAuthMiddleWare({
      cache: mockCache,
      httpLib: mockHttpLib,
      shouldLocalAuth: false,
    });

    authMiddleware(mockRequest, null, (err) => {
      expect(err).not.toBeDefined();
      expect(mockHttpLib.mock.calls).toHaveLength(0);
      expect(mockRequest.kubeToken).toBe('Bearer test-cached-token');
      expect(mockCache.get.mock.calls).toHaveLength(3);
      expect(mockCache.get.mock.calls[0]).toMatchSnapshot();
      done();
    });
  });
});
