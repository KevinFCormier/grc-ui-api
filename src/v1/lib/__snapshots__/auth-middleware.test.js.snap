// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Middleware Caches token on first authorization 1`] = `
Array [
  "ng-for-auth",
  "test-id-token",
]
`;

exports[`Auth Middleware Retrieves token from cache if present 1`] = `
Array [
  "ng-for-auth",
]
`;
